shared/macros-bash.jinja

```diff
@@ -686,3 +686,75 @@ dpkg-query --show --showformat='${db:Status-Status}\n' "{{{ pkgname }}}" 2>/dev/
 rpm --quiet -q "{{{ pkgname }}}"
 {{%- endif -%}}
 {{%- endmacro -%}}
+
+{{#
+  Macro to replace configuration setting in config file or add the configuration setting if
+  it does not exist.
+ 
+  Expects arguments:
+ 
+  config_file:		Configuration file that will be modified
+  key:			Configuration option to change
+  value:		Value of the configuration option to change
+  cce:			The CCE identifier or '@CCENUM@' if no CCE identifier exists
+  format:		The printf-like format string that will be given stripped key and value as arguments,
+ 			so e.g. '%s=%s' will result in key=value subsitution (i.e. without spaces around =)
+ 
+  Optional arugments:
+ 
+  format:		Optional argument to specify the format of how key/value should be
+  			modified/appended in the configuration file. The default is key = value.
+ 
+  Example Call(s):
+ 
+      With default format of 'key = value':
+      {{{ bash_replace_or_append('/etc/sysctl.conf', '^kernel.randomize_va_space', '2', '@CCENUM@') }}}
+ 
+      With custom key/value format:
+      {{{ bash_replace_or_append('/etc/sysconfig/selinux', '^SELINUX=', 'disabled', '@CCENUM@', '%s=%s') }}}
+ 
+      With a variable:
+      {{{ bash_replace_or_append('/etc/sysconfig/selinux', '^SELINUX=', "$var_selinux_state", '@CCENUM@', '%s=%s') }}}
+ 
+#}}
+{{%- macro bash_replace_or_append(config_file, key, value, cce, format='%s = %s') -%}}
+case_insensitive_mode=yes sed_case_insensitive_option='' grep_case_insensitive_option=''
+
+if [ "$case_insensitive_mode" = yes ]; then
+    sed_case_insensitive_option="i"
+    grep_case_insensitive_option="-i"
+fi
+
+# Test if the config_file is a symbolic link. If so, use --follow-symlinks with sed.
+# Otherwise, regular sed command will do.
+sed_command=('sed' '-i')
+if test -L "{{{ config_file }}}"; then
+    sed_command+=('--follow-symlinks')
+fi
+
+# Test that the cce arg is not empty or does not equal @CCENUM@.
+# If @CCENUM@ exists, it means that there is no CCE assigned.
+if [ -n "{{{ cce }}}" ] && [ "{{{ cce }}}" != '@CCENUM@' ]; then
+    cce="{{{ cce }}}"
+else
+    cce="CCE"
+fi
```

Comment:
We are no longer in a function - we know whether `cce` is defined or not at the time when the macro is used, so there can be a cce assignment sub-macro.

Reply:
Hi @matejak , thanks for pointing this out. It's an interesting piece of code. 

The `@CCENUM@` is a string that is replaced at a later phase of the build process by the rule CCE number. It's done by the XSLT template /shared/transforms/xccdf-addremediations.xslt on lines 68 - 81. Notice that in `/build/<product>/fixes/bash` you will see the `@CCENUM@` string, but in the generated SCAP source data stream (`ssg-<product>-ds.xml`) you will see that it got replaced by the rule's CCE number. Or, it's replaced by an empty string if the rule doesn't have a CCE number. For example, in Fedora we don't assign CCEs.

Another observation is that the bash function or now the jinja macro is always called with the third parameter set to `@CCENUM@`, which means that this parameter is probably useless and we could always use `@CCENUM@` in the macros code.

We will probably need to change the logic here to get the desired result in the built SCAP source data stream. The desired result is: If the rule has a CCE number, set the `cce` bash variable to the value of the CCE number. If the rule doesn't have a CCE number, set the `cce` bash variable to string `"CCE"`

Reply:
Commit 2ea91b3 should achieve the desired result; however, it generates weird code in the resulting remediation. If anyone has a more elegant solution, it would be appreciated :eyes: